src/perf/lib/ThroughputServer.cpp

```diff
@@ -0,0 +1,211 @@
+        return Status;
+    }
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Start(
+    _In_ QUIC_EVENT StopEvent
+    ) {
+    QUIC_STATUS Status =
+        Listener.Start(
+            &Address,
+            [](HQUIC Handle, void* Context, QUIC_LISTENER_EVENT* Event) -> QUIC_STATUS {
+                return ((ThroughputServer*)Context)->ListenerCallback(Handle, Event);
+            },
+            this);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+    RefCount = CountHelper{StopEvent};
+    if (NumberOfConnections > 0) {
+        for (uint32_t i = 0; i < NumberOfConnections; i++) {
+            RefCount.AddItem();
+        }
+    } else {
+        //
+        // Add a single item so we can wait on the Count Helper
+        //
+        RefCount.AddItem();
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Wait(
+    _In_ int Timeout
+    ) {
+    if (Timeout > 0) {
+        RefCount.Wait(Timeout);
+    } else {
+        RefCount.WaitForever();
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::ListenerCallback(
+    _In_ HQUIC /*ListenerHandle*/,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    ) {
+    switch (Event->Type) {
+    case QUIC_LISTENER_EVENT_NEW_CONNECTION: {
+        Event->NEW_CONNECTION.SecurityConfig = SecurityConfig;
+        QUIC_CONNECTION_CALLBACK_HANDLER Handler =
+            [](HQUIC Conn, void* Context, QUIC_CONNECTION_EVENT* Event) -> QUIC_STATUS {
+                return ((ThroughputServer*)Context)->
+                    ConnectionCallback(
+                        Conn,
+                        Event);
+            };
+        MsQuic->SetCallbackHandler(
+            Event->NEW_CONNECTION.Connection,
+            (void*)Handler,
+            this);
+        break;
+    }
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::ConnectionCallback(
+    _In_ HQUIC ConnectionHandle,
+    _Inout_ QUIC_CONNECTION_EVENT* Event
+    ) {
+    switch (Event->Type) {
+    case QUIC_CONNECTION_EVENT_CONNECTED:
+        QuicTraceLogInfo(
+            ConnectionEventConnected,
+            "[ conn] Connection Connected (%p).",
+            ConnectionHandle);
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT:
+        QuicTraceLogInfo(
+            ConnectionEventPeerShutdown,
+            "[ conn] Connection Shutdown By Peer (%d) (%p).",
+            Event->SHUTDOWN_INITIATED_BY_TRANSPORT.Status,
+            ConnectionHandle);
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_PEER:
+        QuicTraceLogInfo(
+            ConnectionEventPeerShutdown,
+            "[ conn] Connection Shutdown By Peer (%p).",
+            ConnectionHandle);
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE:
+        QuicTraceLogInfo(
+            ConnectionEventShutdownComplete,
+            "[ conn] Connection Shutdown Complete (%p).",
+            ConnectionHandle);
```

Comment:
Remember that we have logging for all this stuff already in msquic. You shouldn't need to duplicate it. Also, prefix your events with Perf, so they don't collide with any in the core.

Reply:
Ok, so our discussion yesterday should have been just removing the prints rather then adding tracing?

Reply:
All depends on if the log duplicates something in msquic already.